src/inventorymanager.cpp

```diff
@@ -255,82 +255,83 @@ void IMoveAction::apply(InventoryManager *mgr, ServerActiveObject *player, IGame
+		if (from_inv.type == InventoryLocation::DETACHED)
+			src_can_take_count = PLAYER_TO_SA(player)->detached_inventory_AllowMove(
+				*this, try_take_count, player);
+		else if (from_inv.type == InventoryLocation::NODEMETA)
+			src_can_take_count = PLAYER_TO_SA(player)->nodemeta_inventory_AllowMove(
+				*this, try_take_count, player);
+		else
+			src_can_take_count = PLAYER_TO_SA(player)->player_inventory_AllowMove(
+				*this, try_take_count, player);
 		dst_can_put_count = src_can_take_count;
 	} else {
-		// Destination is detached
+		ItemStack restitem;
+		bool swap_expected = list_to->getItem(to_i).count != 0
+			&& !list_to->itemFits(to_i, src_item, &restitem)
+			&& !restitem.empty()
+			&& restitem.count == src_item.count
+			&& !caused_by_move_somewhere;
+
 		if (to_inv.type == InventoryLocation::DETACHED) {
-			ItemStack src_item = list_from->getItem(from_i);
-			src_item.count = try_take_count;
 			dst_can_put_count = PLAYER_TO_SA(player)->detached_inventory_AllowPut(
 				*this, src_item, player);
-		}
-		// Source is detached
-		if (from_inv.type == InventoryLocation::DETACHED) {
-			ItemStack src_item = list_from->getItem(from_i);
-			src_item.count = try_take_count;
-			src_can_take_count = PLAYER_TO_SA(player)->detached_inventory_AllowTake(
+		} else if (to_inv.type == InventoryLocation::NODEMETA) {
+			dst_can_put_count = PLAYER_TO_SA(player)->nodemeta_inventory_AllowPut(
+				*this, src_item, player);
+		} else if (to_inv.type == InventoryLocation::PLAYER) {
+			dst_can_put_count = PLAYER_TO_SA(player)->player_inventory_AllowPut(
 				*this, src_item, player);
 		}
-	}
 
-	/* Query node metadata inventories */
-
-	// Both endpoints are nodemeta
-	// Move occurs in the same nodemeta inventory
-	if (from_inv.type == InventoryLocation::NODEMETA &&
-			from_inv == to_inv) {
-		src_can_take_count = PLAYER_TO_SA(player)->nodemeta_inventory_AllowMove(
-			*this, try_take_count, player);
-		dst_can_put_count = src_can_take_count;
-	} else {
-		// Destination is nodemeta
-		if (to_inv.type == InventoryLocation::NODEMETA) {
-			ItemStack src_item = list_from->getItem(from_i);
-			src_item.count = try_take_count;
-			dst_can_put_count = PLAYER_TO_SA(player)->nodemeta_inventory_AllowPut(
+		if (from_inv.type == InventoryLocation::DETACHED) {
+			src_can_take_count = PLAYER_TO_SA(player)->detached_inventory_AllowTake(
 				*this, src_item, player);
-		}
-		// Source is nodemeta
-		if (from_inv.type == InventoryLocation::NODEMETA) {
-			ItemStack src_item = list_from->getItem(from_i);
-			src_item.count = try_take_count;
+		} else if (from_inv.type == InventoryLocation::NODEMETA) {
 			src_can_take_count = PLAYER_TO_SA(player)->nodemeta_inventory_AllowTake(
 				*this, src_item, player);
+		} else if (from_inv.type == InventoryLocation::PLAYER) {
+			src_can_take_count = PLAYER_TO_SA(player)->player_inventory_AllowTake(
+				*this, src_item, player);
 		}
-	}
 
-	// Query player inventories
+		if (swap_expected) {
```

Comment:
Some comment(s) explaining what this does would be nice.

Reply:
Agreed :+1:

Reply:
Will do as soon as I have time